### DISCUSSION TITLE
RecursionError: maximum recursion depth exceeded when use m2m with jsonable_encoder in fastapi

### DISCUSSION THREAD

**Discussion by wu-clan** [sqlalchemy/sqlalchemy#9094] [Usage Questions] [answered]
This is a simple instance model, which is not complete. The role table has an m2m relationship with the menu table.

```python
class User(MappedAsDataclass, DeclarativeBase):
    __tablename__ = 'sys_user'
    # id_key is a Annotated
    id: Mapped[id_key] = mapped_column(init=False)
    username: Mapped[str] = mapped_column(String(20), unique=True, index=True)
    password: Mapped[str] = mapped_column(String(255))
    email: Mapped[str] = mapped_column(String(50), unique=True, index=True)

    roles: Mapped[List['Role']] = relationship(init=False, secondary='sys_user_role', back_populates='users', cascade='all, delete')

class Role(MappedAsDataclass, DeclarativeBase):
    __tablename__ = 'sys_role'
    id: Mapped[id_key] = mapped_column(init=False)
    name: Mapped[str] = mapped_column(String(32), unique=True)
    description: Mapped[Optional[str]] = mapped_column(LONGTEXT)

    users: Mapped[List['User']] = relationship(init=False, secondary='sys_user_role', back_populates='roles', passive_deletes=True)

sys_user_role = Table(
    'sys_user_role',
    MappedBase.metadata,
    Column('user_id', ForeignKey('sys_user.id', ondelete='CASCADE'), primary_key=True),
    Column('role_id', ForeignKey('sys_role.id', ondelete='CASCADE'), primary_key=True),
)
```
sa select:
```python 
def get_userinfo_by_id(self, db: Session, user_id: int) -> User:
    return db.execute(
        select(self.model)
        .where(self.model.id == user_id)
        # .options(noload(self.model.roles))  # TODO: maximum recursion depth exceeded if select load roles
    ).scalars().first()
```
api：
```python
@user.get('/{username}')
def get_userinfo(username: str, db=Depends(get_db)):)
    current_user = user_service.get_user_info(db=db, username=username)
    return response_base.response_200(
        data=current_user,
        exclude={'password'}
    )
```
Omit details in user_service and response, `user_service`is the logical content, `response_200()`is a custom return，for fastapi，The final return will all go through [jsonable_encoder()](https://github.com/tiangolo/fastapi/blob/master/fastapi/encoders.py) to parse the data, the database query class parsed in this: [dataclasses]( https://github.com/tiangolo/fastapi/blob/5905c3f740c8590f1a370e36b99b760f1ee7b828/fastapi/encoders.py#L73),when parsing `current_user`, ` RecursionError: maximum recursion depth exceeded`, it seems that the associated `roles` parsed the user associated with the third table, then the role associated with the user was parsed again, and the cycle repeats. I'm not sure who to ask about this, whether it's a fastapi problem or a select error on my part, but like in the code, if I use `noload`, it won't generate an error.

**Comment by zzzeek** [comment, marked as answer]
you'd need to look at the jsonable_encoder function, which is likely not able to handle reference cycles.   when you load that relationship, it's very common that User->Role->User cycles will be present in memory, and these must be correctly handled by any serializer.  so not a sqlalchemy issue.

**Comment by wu-clan** [comment]
I will quote this question to `fastapi`，in here https://github.com/tiangolo/fastapi/issues/2972
